eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentPacket.scala

```diff
@@ -87,108 +115,87 @@ object IncomingPaymentPacket {
-                      //  - the blinding point is invalid: the sender or the previous node is buggy or malicious
-                      //  - the encrypted data is invalid: the recipient is buggy or malicious
-                      // TODO: return an unparseable error
-                      Left(InvalidOnionPayload(UInt64(12), 0))
-                    case Success((relayData, nextBlinding)) =>
-                      if (isValidBlindedPayment(relayData, add.amountMsat, add.cltvExpiry, Features.empty)) {
-                        // TODO: If we build routes with several copies of our node at the end to hide the true length of the
-                        // route, then we should add some code here to continue decrypting onions until we reach the final packet.
-                        Right(ChannelRelayPacket(add, PaymentOnion.BlindedChannelRelayData(relayData, add.amountMsat, add.cltvExpiry), next, Some(nextBlinding)))
-                      } else {
-                        // The sender is buggy or malicious, probably trying to probe the blinded route.
-                        // TODO: return an unparseable error
-                        Left(InvalidOnionPayload(UInt64(12), 0))
-                      }
-                  }
-                case None =>
-                  // The sender is trying to use route blinding, but we didn't receive the blinding point used to derive
-                  // the decryption key. The sender or the previous peer is buggy or malicious.
-                  // TODO: return an unparseable error
-                  Left(InvalidOnionPayload(UInt64(12), 0))
-              }
+    decryptOnion(add.paymentHash, outerOnionDecryptionKey, add.onionRoutingPacket).flatMap {
+      case DecodedOnionPacket(payload, Some(nextPacket)) =>
+        payload.get[OnionPaymentPayloadTlv.EncryptedRecipientData] match {
+          case Some(OnionPaymentPayloadTlv.EncryptedRecipientData(encryptedRecipientData)) =>
+            decryptEncryptedRecipientData(add, privateKey, payload, encryptedRecipientData).flatMap {
+              case DecodedEncryptedRecipientData(blindedPayload, nextBlinding) =>
+                validateBlindedChannelRelayPayload(add, payload, blindedPayload, nextBlinding, nextPacket)
             }
-          case _ if add.blinding_opt.isDefined => Left(InvalidOnionPayload(UInt64(12), 0))
-          // NB: we don't validate the ChannelRelayPacket here because its fees and cltv depend on what channel we'll choose to use.
-          case payload: PaymentOnion.ChannelRelayTlvPayload => Right(ChannelRelayPacket(add, payload, next, None))
+          case None if add.blinding_opt.isDefined => Left(InvalidOnionPayload(UInt64(12), 0))
+          case None => PaymentOnion.ChannelRelayPayload.validate(payload).left.map(_.failureMessage).map {
+            // NB: we don't validate the ChannelRelayPacket here because its fees and cltv depend on what channel we'll choose to use.
```

Comment:
In other words there is nothing to validate right?

Reply:
I clarified the comment in https://github.com/ACINQ/eclair/pull/2414/commits/318e59e425ebe19877db1800523aa31e788995f8, let me know if that's better

Reply:
I don't see your correction?

Reply:
That's strange, I do see it in the commit I'm linking...is this a github caching issue?

Reply:
I think the validation for other messages are about the integrity and self-constency of the payload, there just isn't any validation to do for the `ChannelRelayPacket`. The fact that some other checks are happening later in the relayer is another matter, there may be other checks for other types of packets too?

Reply:
Yes that's true, it's a good argument for just removing this comment entirely, I'll do that.

Reply:
Removed in https://github.com/ACINQ/eclair/pull/2414/commits/2382b63691cbe853257f264398a03f51429fd56d